tracee-ebpf/tracee/tracee.go

```diff
@@ -2086,3 +2104,228 @@ func readArgFromBuff(dataBuff io.Reader) (argTag, interface{}, error) {
+
+func (c *Containers) _CheckContainer(contid string) bool {
+	// check if container exists
+	_, ok := c.mapContIds[contid]
+	if !ok {
+		return false
+	}
+	return true
+}
+
+func (c *Containers) GetContainers() []string {
+	// list existing containers
+	var conts []string
+	for k, _ := range c.mapContIds {
+		conts = append(conts, k)
+	}
+	if len(conts) == 0 {
+		return nil
+	}
+	return conts
+}
+
+func (c *Containers) _AddPid(contid string, pid string) {
+	// add pid to container
+	c._AddContainer(contid)
+	c.mapContIds[contid] = InsertSortUniq(c.mapContIds[contid], pid)
+}
+
+func (c *Containers) _DelPid(contid string, pid string) {
+	// del pid from container
+	if c._CheckContainer(contid) == true {
+		c.mapContIds[contid] = DeleteSortUniq(c.mapContIds[contid], pid)
+	}
+}
+
+func (c *Containers) GetPids(contid string) []string {
+	// list pids from existing container
+	if c._CheckContainer(contid) == false {
+		return nil
+	}
+	return c.mapContIds[contid]
+}
+
+// Populate info
+
+func (c *Containers) Populate() error {
+	var err error
+
+	err = c._ProcMountsCgroups()
+	if err != nil {
+		return err
+	}
+	err = c._PopulateDocker()
+	if err != nil {
+		return err
+	}
+	err = c._PopulatePodman()
+	if err != nil {
+		return err
+	}
+
+	return err
+}
+
+func (c *Containers) Get() map[string][]string {
+	// easify range for map[string][]string
+	return c.mapContIds
+}
+
+func (c *Containers) _PopulateProcWalk(basedir string, re *regexp.Regexp) error {
+
+	err := filepath.Walk(basedir,
+		func(fn string, fi os.FileInfo, err error) error {
+			if err != nil {
+				return nil // no error in procfs reads
+			}
+			if re.MatchString(fn) == false {
+				return nil // no error if no match
+			}
+			dir := string(filepath.Dir(fn))
+			pos := strings.LastIndex(dir, "/")
+			contid := dir[pos+1:]
+
+			// should transform parent directory ?
+			res := re.FindStringSubmatch(fn)
+			names := re.SubexpNames()
+			for i, o := range res {
+				if i == 1 && names[i] == "id" {
+					contid = o
+				}
+			}
+			// container id transformed based on regex grp "id"
+			c._AddContainer(contid)
+
+			buf, e := ioutil.ReadFile(fn)
+			if e != nil {
+				return e // error if can't read pids file
+			}
+			for _, pid := range strings.Split(string(buf), "\n") {
+				if len(pid) > 0 {
+					c._AddPid(contid, pid)
+				}
+			}
+			return nil
+		})
+
+	return err
+}
+
+func (c *Containers) _PopulateDocker() error {
+	if c.cgroupv1 {
+		// path: <cgroupv1mp>/<random>/docker/<id>/tasks
+		re := regexp.MustCompile(c.cgroupv1mp + ".*docker.*(?P<id>.{64})/tasks$")
+		err := c._PopulateProcWalk(c.cgroupv1mp, re)
+		if err != nil {
+			return err
+		}
+	}
+	if c.cgroupv2 {
+		// path: <cgroupv2mp>/<random>/docker-<id>.scope/cgroup.procs
+		re := regexp.MustCompile(c.cgroupv2mp + `.*docker.*(?P<id>.{64})\.scope/cgroup.procs$`)
+		err := c._PopulateProcWalk(c.cgroupv2mp, re)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (c *Containers) _PopulatePodman() error {
```

Comment:
This function and _PopulateDocker() are the same but with a different regex.
We can have one function that gets the regex string as an argument and avoid code duplication for different container runtimes.